### ISSUE TITLE
Engine Build Script rewrite #2: Nightly builds

### ISSUE THREAD

**Issue by wdanilo** [enso-org/enso#4192]
This task is automatically imported from the old Task Issue Board and it was originally created by Wojciech Daniło.
Original issue is [here](https://www.pivotaltracker.com/story/show/180302838).
<hr />

## Description
This is the second step necessary for the new CI infrastructure. See the epic description for details.

## Resources
This involves creating a functional equivalent to the `nightly.yml` GitHub actions workflow.

## Blockers:
[#180302836](https://www.pivotaltracker.com/story/show/#180302836) resolved

## Comments:
**Michał Urbańczyk** reports a new **STANDUP** for yesterday (2021-12-28):

**Today:** Split the current repository for CI into two: one for deployment logic (that is private) and the other one for prototyping the build script (public). This will give us unlimited CI time for testing the build script and will make us less reliant on self-hosted macOS runners.

Started writing new version generation logic. It should be finished by 2022-01-12.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180302838). Continue rewriting the enso nightly release logic. Setup CI on the provisional build repo. (Enso Bot - Dec 29, 2021)
<hr />
**Michał Urbańczyk** reports a new **STANDUP** for yesterday (2021-12-29):

**Today:** Rewriting the nigthly release build. Got to the point where new version numbers are generated, propagated upon SBT and packages build. Refactored the runner deployment code to support organization-level runners. It should be finished by 2022-01-12.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180302838). Continue rewriting the enso nightly release logic. (Enso Bot - Dec 30, 2021)
<hr />
**Michał Urbańczyk** reports a new **STANDUP** for today (2021-12-30):

**Today:** Rewriting the nigthly release build. Integrating nightly release and the normal build code. It should be finished by 2022-01-12.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180302838). Continue rewriting the enso nightly release logic. Particularly, look into creating GH releases with assets from code. (Enso Bot - Dec 30, 2021)
<hr />
**Michał Urbańczyk** reports a new **🔴 DELAY** for today (2022-01-17):

**Summary:** There is **14 days** delay in implementation of the [Engine Build Script rewrite #2: Nightly builds (#180302838) task](https://www.pivotaltracker.com/story/show/180302838).
It **will cause 14 days delay** for the delivery of this sprint.

**Delay Cause:** Not really lengthening this task estimate, just offsetting for the fact that it was suspended and is now being resumed. The delay is mostly my time off and weekends. Rest is time spent on #180855729. (Enso Bot - Jan 18, 2022)
<hr />
**Michał Urbańczyk** reports a new **STANDUP** for today (2022-01-17):

**Today:** I've returned to the new nightly build script. Iteration time is awful.

Noticed some issues with package verification script, though I'll leave that for later. 
Started rewriting the packaging script. This should go quick, as it is mostly about moving things around and packing.

While things were building, I've also done some experiments about improving build-times. Still no "quick" solution in the sight. It should be finished by 2022-01-26.

**Next Day:** Next day I will be working on the [##180302838 task](https://www.pivotaltracker.com/story/show/180302838). Investigate the issue with package verification script.
Finish the build/package step.

Further in pipeline (not tomorrow) are: parsing changelog, creating & uploading release (already done prototype in the meantime). Finally, there's AWS integration for release upload. (And, after all is functional, deployment to the main repo). (Enso Bot - Jan 18, 2022)
<hr />
**Michał Urbańczyk** reports a new **STANDUP** for today (2022-01-18):

**Today:** Mostly done packaging & bundling. Not as quick job as I thought. Haven't checked the verification script (but likely I'll want to rewrite it anyway).
Had an informative talk with Radek about CI. It should be finished by 2022-01-26.

**Next Day:** Next day I will be working on the [##180302838 task](https://www.pivotaltracker.com/story/show/180302838). Finish the nightly release creation and asset upload. Setup the thing on a provisional repository as CI job. (Enso Bot - Jan 19, 2022)
<hr />
**Michał Urbańczyk** reports a new **🔴 DELAY** for yesterday (2022-01-19):

**Summary:** There is **1 days** delay in implementation of the [Engine Build Script rewrite #2: Nightly builds (#180302838) task](https://www.pivotaltracker.com/story/show/180302838).
It **will cause 1 days delay** for the delivery of this sprint.

**Delay Cause:** I'm sorry, there is no  extraordinary reason. Progress is steady, things just seem to take longer overall than I initially estimated. Thus I wanted to give heads-up and report possible delay early. (Enso Bot - Jan 20, 2022)
<hr />
**Michał Urbańczyk** reports a new **STANDUP** for yesterday (2022-01-19):

**Today:** Finishing and deploying packaging scripts.  Not done as I expected. It should be finished by 2022-01-27.

**Next Day:** Next day I will be working on the [##180302838 task](https://www.pivotaltracker.com/story/show/180302838). Continue the current stage of work — finish the nightly release creation and asset upload. Setup the thing on a provisional repository as CI job. (Enso Bot - Jan 20, 2022)
<hr />
**Michał Urbańczyk** reports a new **STANDUP** for the last Friday (2022-01-21):

**Today:** Finished the asset creation and upload. Refactored structure so there is a separate release preparation step (so different runs won't end up competing for the same version). First releases with complete set of assets were created semi-automatically. Started work on introducing the new version scheme and changelog parsing.

Tested the generated assets contents. Some minor issues remain:
* graalvm distribution has broken symlinks, leading to unnecessary size increase;
* missing IR caches in the standard library distribution;
* one pass to rename things.

Introduced some optimizations to the package building process, cutting the build times to 15–20 min on self-hosted Windows and Linux runners. They might increase by a few minutes, as I need to restore IR cache compilation. Still, its about twice as fast as old nightly engine CI. It should be finished by 2022-01-27.

**Next Day:** Next day I will be working on the [##180302838 task](https://www.pivotaltracker.com/story/show/180302838). Polish the release itself — manage draft-publish stage, include relevant excerpt from the changelog. Work on addressing the discovered package maladies.

**Additional Information:** This standup covers Thursday and Friday. Sorry for that, work somewhat sprawled across the days. (Enso Bot - Jan 24, 2022)
<hr />
**Michał Urbańczyk** reports a new **STANDUP** for yesterday (2022-01-24):

**Today:** Changelog parsing and generating version numbers. Finishing the draft-upload-publish scheme. It should be finished by 2022-01-27.

**Next Day:** Next day I will be working on the [##180302838 task](https://www.pivotaltracker.com/story/show/180302838). Refactor to unify the nightly release and on-PR logic. Work on issues within packages. (Enso Bot - Jan 26, 2022)
<hr />
**Michał Urbańczyk** reports a new **STANDUP** for yesterday (2022-01-25):

**Today:** Fixed the issue with standard library IR caches not being part of the package. Improvements to CLI argument handling code. Tested using the updated script to build non-nightly packages, fixed found issues. Preparing for a weekly epic call. It should be finished by 2022-01-27.

**Next Day:** Next day I will be working on the [##180302838 task](https://www.pivotaltracker.com/story/show/180302838). Fix the issue with GraalVM distribution symlinks not being preserved, leading to overly heavy non-Windows packages.
One more pass to check the generated packages and compare them with ones produced by the old scripts. (Enso Bot - Jan 26, 2022)
<hr />
**Michał Urbańczyk** reports a new **STANDUP** for yesterday (2022-01-26):

**Today:** Fixed the symlink handling when packaging the Graal runtime. Ended up involving a bigger refactoring than I expected, but results are nice. Minor testing CI tweaks. 
It seems that the new script is now able to correctly create, upload and publish a nightly release to a GitHub.
Did some early prototyping for the artifacts API wrappers, managed to actually upload something. It should be finished by 2022-01-27.

**Next Day:** Next day I will be working on the [##180302838 task](https://www.pivotaltracker.com/story/show/180302838). AWS edition updating for the published release. Addressing remaining issues. (Enso Bot - Jan 27, 2022)
<hr />
**Michał Urbańczyk** reports a new **🔴 DELAY** for yesterday (2022-01-27):

**Summary:** There is **5 days** delay in implementation of the [Engine Build Script rewrite #2: Nightly builds (#180302838) task](https://www.pivotaltracker.com/story/show/180302838).
It **will cause 5 days delay** for the delivery of this sprint.

Not really that much of a delay, I've decided to work on the Artifact API wrappers earlier and use them when delivering this task, so I don't have to rewrite this later. Work-day delay is around a day.

**Delay Cause:** Given decision that we want to investigate Artifact API opportunity and promising results of my early prototype, delaying finishing this task makes sense: the final part (being AWS editions upload) will benefit from the endevour. That part is more of rescheduling than delay.
"Actual" delay is about:
* SBT and cmd.exe not playing nicely with quotes — leading me into an investigation of faux failures in script that verifies license notices, instead of the actual failures;
* one engine test that needs fixing (it does not like the version scheme), I've asked Radek for help with this. (Enso Bot - Jan 28, 2022)
<hr />
**Michał Urbańczyk** reports a new **STANDUP** for yesterday (2022-01-27):

**Today:** Fixed the issue with the script verifying the licenses. Implemented the provisional AWS integration for managing nightly releases, it will be integrated when Artifact API is functional. It should be finished by 2022-02-01.

**Next Day:** Next day I will be working on the [##180302838 task](https://www.pivotaltracker.com/story/show/180302838). Work on Artifact API. Capability of being able to upload artifacts directly from Rust. Intercepting relevang pieces of environment. Test integration with the current build script. (Enso Bot - Jan 28, 2022)
<hr />
**Michał Urbańczyk** reports a new **STANDUP** for the last Friday (2022-01-28):

**Today:** Artifact API — mostly done upload file, though I'm going for a slightly simpler approach than in the official GH toolkit code. 
Enabling access to artifact API by capturing relevant environment variables. Some reworking of the yaml workflow files. It should be finished by 2022-02-01.

**Next Day:** Next day I will be working on the [##180302838 task](https://www.pivotaltracker.com/story/show/180302838). Finish the artifact upload. Start work on artifact download. Fix the issue with missing components in the project manager bundle. (Enso Bot - Jan 31, 2022)
<hr />
**Michał Urbańczyk** reports a new **STANDUP** for today (2022-01-31):

**Today:** Worked on the Artifact API. Spend more time than expected. Got upload to work properly with big files and upload filesystem filetrees. Support asynchronous file discoverers. Still needs cleanups. It should be finished by 2022-02-01.

**Next Day:** Next day I will be working on the [##180302838 task](https://www.pivotaltracker.com/story/show/180302838). Focus on the remaining nightly work. Deploy it to the Enso-Staging repository. Advance the Artifact API. (Enso Bot - Feb 1, 2022)
<hr />
**Michał Urbańczyk** reports a new **🔴 DELAY** for today (2022-02-01):

**Summary:** There is **1 days** delay in implementation of the [Engine Build Script rewrite #2: Nightly builds (#180302838) task](https://www.pivotaltracker.com/story/show/180302838).
It **will cause 1 days delay** for the delivery of this sprint.

**Delay Cause:** CI issues caused by the GH runners update failure — spent some figuring this out and redeploying runners. Artifacts API took a little more time than it should. (Enso Bot - Feb 2, 2022)
<hr />
**Michał Urbańczyk** reports a new **STANDUP** for today (2022-02-01):

**Today:** Did some finalizing work on nightly script, though did not manage to complete it. Debugging and fixing the issues with self-hosted Windows runners, that entered an infinite "download update and restart to install" loop. Had a call about running benchmarks from CI and a weekly Epic call. It should be finished by 2022-02-02.

**Next Day:** Next day I will be working on the [##180302838 task](https://www.pivotaltracker.com/story/show/180302838). Focus on the remaining nightly work. Deploy it to the Enso-Staging repository. Advance the Artifact API. (Enso Bot - Feb 2, 2022)
<hr />
**Michał Urbańczyk** reports a new **🔴 DELAY** for yesterday (2022-02-02):

**Summary:** There is **2 days** delay in implementation of the [Engine Build Script rewrite #2: Nightly builds (#180302838) task](https://www.pivotaltracker.com/story/show/180302838).
It **will cause 2 days delay** for the delivery of this sprint.

**Delay Cause:** Deploying the updated script to `enso` revealed new issues with dev builds — the fix for engine tests proved ineffective. Apparently, they required version number to be between 0.1.0 and 0.999.0 — which neither 0.0.0-dev nor 2021.1.1. do not fullfill. Additionally, after reviewing the branch together, we found an issue with the approach taken and figured out a better way to deal with stdlib directory names that can be provided.

I hope to receive help from Radek with finishing this version scheme follow-up fixes.

I've also got tricked by the git and caused trouble for myself by failing to update some files which were in the ignore list despite being versioned. (Enso Bot - Feb 3, 2022)
<hr />
**Michał Urbańczyk** reports a new **STANDUP** for yesterday (2022-02-02):

**Today:** Cleanups in the build script and deploying an updated version. Wasted much time on changing the stdlib version — the script for this seemingly works on windows, but somehow breaks newlines, causing test to fail at later point.
Deployed updated test script, resolved issues with handling both nightly and dev builds from the same script. It should be finished by 2022-02-04.

**Next Day:** Next day I will be working on the [##180302838 task](https://www.pivotaltracker.com/story/show/180302838). <no point in filling this, as I'm reporting both standups at the same time> (Enso Bot - Feb 3, 2022)
<hr />
**Michał Urbańczyk** reports a new **STANDUP** for today (2022-02-03):

**Today:** Had a call with Radek and discussed issues discovered following the script update rollout and integration PR to the main repository.
I've fixed the remaining (upper bound) issue with version number. 
Nightly builds are provisionally released through `ci-build` repository.
PR should be mergable tomorrow, with Radek help.

In the meantime, I've started GUI script rewrite. It should be finished by 2022-02-04.

**Next Day:** Next day I will be working on the [##180302838 task](https://www.pivotaltracker.com/story/show/180302838). Merge the PR and have this thing finally done. While blocked or done, work towards the next task (IDE build). (Enso Bot - Feb 3, 2022)
<hr />
**Michał Urbańczyk** reports a new **🔴 DELAY** for today (2022-02-07):

**Summary:** There is **1 days** delay in implementation of the [Engine Build Script rewrite #2: Nightly builds (#180302838) task](https://www.pivotaltracker.com/story/show/180302838).
It **will cause 1 days delay** for the delivery of this sprint.

**Delay Cause:** The seemingly endless series of issues following the version scheme change persisted. Only today the nightly preparations PR was merged into Enso. However, this means I can start setting up the experimental deployment on tomorrow.

The good news is that the PR finally landed, so no more surprises are expected there.
The other good news is that we actually were able to finally get rid of the need to change directory names in sources working copy with the special script to have consistent versioning. (Enso Bot - Feb 7, 2022)
<hr />
**Michał Urbańczyk** reports a new **🔴 DELAY** for today (2022-02-07):

**Summary:** There is **4 days** delay in implementation of the [Engine Build Script rewrite #2: Nightly builds (#180302838) task](https://www.pivotaltracker.com/story/show/180302838).
It **will cause 4 days delay** for the delivery of this sprint.

**Delay Cause:** I forgot about the weekend. Delay was actually 2 days. (Enso Bot - Feb 7, 2022)
<hr />
**Michał Urbańczyk** reports a new **STANDUP** for today (2022-02-07):

**Today:** Developing, debugging, discussing the nightly preparations PR. Finally merged today, with significant help from Radek.
While being blocked by the Engine build issues, I was investigating how the IDE build scripts are structured and I had a call with Wojciech regarding the design for the next iteration of GUI build scripts.

Also, I was looking into the CI issue reported by Edward, where persistence of Engine cache directory under HOME caused builds to fail. After discussion with Radek as to why is this happening, I've deployed a fix. It should be finished by 2022-02-08.

**Next Day:** Next day I will be working on the [##180302838 task](https://www.pivotaltracker.com/story/show/180302838). Deploy nightly builds to engine staging. Add missing pieces of Artifacts API. (Research in the meantime showed that listing artifacts for the run through official Actions API does not work for us in this case.)

**Additional Information:** Covers also Friday. Sorry, for that, first I was waiting, because I've done some work during weekend, then today I waited to see if the PR situation is resolved. (Enso Bot - Feb 7, 2022)
<hr />
**Michał Urbańczyk** reports a new **🔴 DELAY** for yesterday (2022-02-08):

**Summary:** There is **1 days** delay in implementation of the [Engine Build Script rewrite #2: Nightly builds (#180302838) task](https://www.pivotaltracker.com/story/show/180302838).
It **will cause 1 days delay** for the delivery of this sprint.

**Delay Cause:** Deploying the new nightly build script to the staging repository shown that there is some issue when passing the AWS authorization info from GH Actions secrets. Needs investigation and fix. (Enso Bot - Feb 9, 2022)
<hr />
**Michał Urbańczyk** reports a new **STANDUP** for yesterday (2022-02-08):

**Today:** Debugging the AWS authentication issue that appeared on enso-staging nightly release workflow deployment. Prepared presentation for the weekly call. It should be finished by 2022-02-09.

**Next Day:** Next day I will be working on the [##180302838 task](https://www.pivotaltracker.com/story/show/180302838). Continue debugging the issue. (Enso Bot - Feb 9, 2022)
<hr />
**Michał Urbańczyk** reports a new **STANDUP** for today (2022-02-09):

**Today:** Finally resolved the AWS issue. Surprisingly, it seems that there was a trailing newline hidden in our secrets. Ugh…
Looked into the issue with CI builds following the Graal version bump. Implemented the mechanism for downloading the appropriate version but more work is needed. It should be finished by 2022-02-09.

**Next Day:** Next day I will be working on the [##181207398 task](https://www.pivotaltracker.com/story/show/181207398). Fix the current CI issues, primarily the Graal version bump aftermath. (Enso Bot - Feb 9, 2022)
<hr />
